File: backend-java/monolith/src/main/java/com/freelms/lms/auth/service/UserService.java
package com.freelms.lms.auth.service;

import com.freelms.lms.auth.dto.UpdateUserRequest;
import com.freelms.lms.auth.dto.UserDto;
import com.freelms.lms.auth.entity.User;
import com.freelms.lms.auth.mapper.UserMapper;
import com.freelms.lms.auth.repository.UserRepository;
import com.freelms.lms.common.dto.PagedResponse;
import com.freelms.lms.common.enums.UserRole;
import com.freelms.lms.common.exception.ResourceNotFoundException;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Slf4j
@Service
@RequiredArgsConstructor
public class UserService {

    private final UserRepository userRepository;
    private final UserMapper userMapper;

    @Cacheable(value = "users", key = "#id")
    @Transactional(readOnly = true)
    public UserDto getUserById(Long id) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("User", "id", id));
        return userMapper.toDto(user);
    }

    @Transactional(readOnly = true)
    public UserDto getUserByEmail(String email) {
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new ResourceNotFoundException("User", "email", email));
        return userMapper.toDto(user);
    }

    @Transactional(readOnly = true)
    public PagedResponse<UserDto> getAllUsers(Pageable pageable) {
        Page<User> users = userRepository.findAll(pageable);
        List<UserDto> dtos = userMapper.toDtoList(users.getContent());
        return PagedResponse.of(users, dtos);
    }

    @Transactional(readOnly = true)
    public PagedResponse<UserDto> getUsersByRole(UserRole role, Pageable pageable) {
        Page<User> users = userRepository.findByRole(role, pageable);
        List<UserDto> dtos = userMapper.toDtoList(users.getContent());
        return PagedResponse.of(users, dtos);
    }

    @Transactional(readOnly = true)
    public PagedResponse<UserDto> getUsersByOrganization(Long organizationId, Pageable pageable) {
        Page<User> users = userRepository.findByOrganizationId(organizationId, pageable);
        List<UserDto> dtos = userMapper.toDtoList(users.getContent());
        return PagedResponse.of(users, dtos);
    }

    @Transactional(readOnly = true)
    public PagedResponse<UserDto> searchUsers(String query, Pageable pageable) {
        Page<User> users = userRepository.searchUsers(query, pageable);
        List<UserDto> dtos = userMapper.toDtoList(users.getContent());
        return PagedResponse.of(users, dtos);
    }

    @CacheEvict(value = "users", key = "#id")
    @Transactional
    public UserDto updateUser(Long id, UpdateUserRequest request) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("User", "id", id));

        if (request.getFirstName() != null) {
            user.setFirstName(request.getFirstName());
        }
        if (request.getLastName() != null) {
            user.setLastName(request.getLastName());
        }
        if (request.getPhoneNumber() != null) {
            user.setPhoneNumber(request.getPhoneNumber());
        }
        if (request.getAvatarUrl() != null) {
            user.setAvatarUrl(request.getAvatarUrl());
        }
        if (request.getBio() != null) {
            user.setBio(request.getBio());
        }
        if (request.getTimezone() != null) {
            user.setTimezone(request.getTimezone());
        }
        if (request.getLanguage() != null) {
            user.setLanguage(request.getLanguage());
        }

        user = userRepository.save(user);
        log.info("User updated: {}", id);
        return userMapper.toDto(user);
    }

    @CacheEvict(value = "users", key = "#id")
    @Transactional
    public void deactivateUser(Long id) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("User", "id", id));
        user.setActive(false);
        userRepository.save(user);
        log.info("User deactivated: {}", id);
    }

    @CacheEvict(value = "users", key = "#id")
    @Transactional
    public void activateUser(Long id) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("User", "id", id));
        user.setActive(true);
        userRepository.save(user);
        log.info("User activated: {}", id);
    }

    @CacheEvict(value = "users", key = "#id")
    @Transactional
    public void updateUserRole(Long id, UserRole newRole) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("User", "id", id));
        user.setRole(newRole);
        userRepository.save(user);
        log.info("User role updated: {} -> {}", id, newRole);
    }

    @Transactional
    public void addPoints(Long userId, int points) {
        userRepository.addPoints(userId, points);
        log.info("Added {} points to user: {}", points, userId);
    }

    @Transactional(readOnly = true)
    public List<UserDto> getUsersByIds(List<Long> ids) {
        List<User> users = userRepository.findByIdIn(ids);
        return userMapper.toDtoList(users);
    }

    @Transactional(readOnly = true)
    public long countUsersByRole(UserRole role) {
        return userRepository.countByRole(role);
    }

    @Transactional(readOnly = true)
    public PagedResponse<UserDto> getTopUsers(UserRole role, Pageable pageable) {
        Page<User> users = userRepository.findTopUsersByRole(role, pageable);
        List<UserDto> dtos = userMapper.toDtoList(users.getContent());
        return PagedResponse.of(users, dtos);
    }
}

File: frontend/src/app/core/services/user.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { environment } from '@environments/environment';
import {
  ApiResponse,
  PagedResponse,
  User,
  UserRole,
  UpdateUserRequest
} from '../models';

@Injectable({
  providedIn: 'root'
})
export class UserService {
  private readonly apiUrl = `${environment.apiUrl}/users`;

  constructor(private http: HttpClient) {}

  getUserById(id: number): Observable<User> {
    return this.http.get<ApiResponse<User>>(`${this.apiUrl}/${id}`)
      .pipe(map(res => res.data));
  }

  getAllUsers(page = 0, size = 20): Observable<PagedResponse<User>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());

    return this.http.get<ApiResponse<PagedResponse<User>>>(this.apiUrl, { params })
      .pipe(map(res => res.data));
  }

  getUsersByRole(role: UserRole, page = 0, size = 20): Observable<PagedResponse<User>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());

    return this.http.get<ApiResponse<PagedResponse<User>>>(`${this.apiUrl}/role/${role}`, { params })
      .pipe(map(res => res.data));
  }

  getUsersByOrganization(organizationId: number, page = 0, size = 20): Observable<PagedResponse<User>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());

    return this.http.get<ApiResponse<PagedResponse<User>>>(`${this.apiUrl}/organization/${organizationId}`, { params })
      .pipe(map(res => res.data));
  }

  searchUsers(query: string, page = 0, size = 20): Observable<PagedResponse<User>> {
    const params = new HttpParams()
      .set('query', query)
      .set('page', page.toString())
      .set('size', size.toString());

    return this.http.get<ApiResponse<PagedResponse<User>>>(`${this.apiUrl}/search`, { params })
      .pipe(map(res => res.data));
  }

  updateCurrentUser(request: UpdateUserRequest): Observable<User> {
    return this.http.put<ApiResponse<User>>(`${this.apiUrl}/me`, request)
      .pipe(map(res => res.data));
  }

  updateUser(id: number, request: UpdateUserRequest): Observable<User> {
    return this.http.put<ApiResponse<User>>(`${this.apiUrl}/${id}`, request)
      .pipe(map(res => res.data));
  }

  updateUserRole(id: number, role: UserRole): Observable<void> {
    const params = new HttpParams().set('role', role);
    return this.http.put<ApiResponse<void>>(`${this.apiUrl}/${id}/role`, null, { params })
      .pipe(map(() => undefined));
  }

  deactivateUser(id: number): Observable<void> {
    return this.http.post<ApiResponse<void>>(`${this.apiUrl}/${id}/deactivate`, {})
      .pipe(map(() => undefined));
  }

  activateUser(id: number): Observable<void> {
    return this.http.post<ApiResponse<void>>(`${this.apiUrl}/${id}/activate`, {})
      .pipe(map(() => undefined));
  }

  getLeaderboard(role: UserRole = UserRole.STUDENT, size = 10): Observable<PagedResponse<User>> {
    const params = new HttpParams()
      .set('role', role)
      .set('size', size.toString());

    return this.http.get<ApiResponse<PagedResponse<User>>>(`${this.apiUrl}/leaderboard`, { params })
      .pipe(map(res => res.data));
  }
}
